Cluster sys version

nicholastmosher opened the pull request:
Add `chart_version` to `SysConfig`. Also puts back `build.rs` in all of the crates that need VERSION file in order to prevent problems with UNDEFINED version on certain configurations.

Bumps fluvio-cluster to `0.7.0` because the refactoring PR that was merged earlier had breaking changes.
Bumps fluvio platform VERSION to `0.7.0-alpha.1` due to the protocol changes a little while ago

sehz commented on .gitignore:
```diff
@@ -5,3 +5,4 @@ target/
 target-docker
 .vscode/tasks.json
 .idea/
+VERSION
```
This make it not check-in the file at the top.

nicholastmosher replied:
git is smart enough to continue tracking changes to files that match .gitignore rules if those files are already committed in the repository

sehz commented on src/cluster/build.rs:
```diff
@@ -0,0 +1,4 @@
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    std::fs::copy("../../VERSION", "./src/VERSION")?;
```
this may fail in the `cargo publish`

nicholastmosher replied:
Hmm I thought this was the solution we used originally and that it worked? Or am I mistaken? I have never used `cargo publish` so I don't know how picky it is and I can't really test it out

sehz replied:
This will fail because VERSION doesn't exists

nicholastmosher replied:
Hmm can we make a custom publish flow that just does `cargo check && cargo publish ARGS`? I think that running cargo check should trigger the VERSION copy and place the file in the crate directory, so it should get zipped up by cargo publish. What do you think of that idea?

nicholastmosher replied:
Actually, since I have made the `copy` infallible, I think that we can make a manual publish flow work because you can just run the `cargo check` or `cargo build` locally yourself before running `cargo publish` and it should work just fine

sehz commented on src/sc/build.rs:
```diff
@@ -0,0 +1,4 @@
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    std::fs::copy("../../VERSION", "./src/VERSION")?;
```
allow failure

nicholastmosher commented:
There were actually some underlying bugs where CLI options were not being propagated into the installer configuration. I have fixed those issues

nicholastmosher commented:
Once we figure out a solution for VERSION, I can add back the version bump to 0.7.0-alpha.1 and we can merge

nicholastmosher commented:
Newest changes:

- `chart_version` is now a required field for both `SysConfig` and `ClusterConfig`
- `fluvio-cli` passes VERSION to `fluvio-cluster
- Fixed broken doc tests
- Added tests for required config options

sehz commented on src/cli/build.rs:
```diff
@@ -2,6 +2,9 @@ use std::process::Command;
 
 fn main() {
+    // Copy VERSION file. Do not fail e.g. when built via `cargo publish`
+    let _ = std::fs::copy("../../VERSION", "./src/VERSION");
```
maybe we put it back old way of referring VERSION?  That way, we don't need to maintain `build.rs`

nicholastmosher replied:
The CLI already has a build.rs that we need to maintain. I think that this is the simplest way to handle our VERSION needs, I really (_really_) disliked having feature flags that set VERSION="UNDEFINED".

sehz replied:
Can you add `rerun-if-changed` this to `build.rs` ?  https://stackoverflow.com/questions/49077147/how-can-i-force-build-rs-to-run-again-without-cleaning-my-whole-project

sehz commented on src/client/src/consumer.rs:
```diff
@@ -37,7 +37,7 @@ use crate::spu::SpuPool;
 ///
```
Can you move this as separate PR since this is not related to this purpose of this PR

sehz commented on src/cluster/src/main_disabled.rs:
```diff
@@ -0,0 +1,12 @@
+// use fluvio_future::task::run_block_on;
+
+// fn main() -> color_eyre::Result<()> {
+//     color_eyre::config::HookBuilder::blank()
```
remove if not used

sehz commented on src/cluster/src/start/k8.rs:
````diff
@@ -49,103 +51,27 @@ static MAX_SC_NETWORK_LOOP: Lazy<u16> = Lazy::new(|| {
-    /// # use fluvio_cluster::ClusterInstaller;
-    /// let installer = ClusterInstaller::new()
+    /// ```
+    /// # use fluvio_cluster::{ClusterConfig, ClusterConfigBuilder, ClusterError};
````
Suggest simplifying doc.   Instead of duplicating examples.  Just have a single example.   Then each setting will just detail what it does.

nicholastmosher replied:
The reason for having individual docs is so that the doc tests will fail if we change the interface. I actually want to open another issue because it turns out our CI has not actually been running doc tests and our docs were out of date.

sehz replied:
Still like to have a just a single example

sehz commented on src/cluster/src/start/k8.rs:
```diff
@@ -49,103 +51,27 @@ static MAX_SC_NETWORK_LOOP: Lazy<u16> = Lazy::new(|| {
-    }
-
+#[derive(Builder, Debug)]
+#[builder(build_fn(skip), setter(prefix = "with"))]
```
I would to remove `with` prefix.  It seems to have more verbose.

nicholastmosher replied:
This gives us the same API we already had. Our builder functions already used names such as `with_chart_version`. This is just telling the `derive_builder` crate how to take a field name e.g. `chart_version` and produce a method for us, `with_chart_version`

sehz commented on src/cluster/src/start/k8.rs:
```diff
@@ -560,94 +632,47 @@ impl ClusterInstallerBuilder {
 pub struct ClusterInstaller {
     /// Configuration options for this installation
-    config: ClusterInstallerBuilder,
+    config: ClusterConfig,
```
Do we need another layer of abstraction?  I would prefer we don't make changes to how config works until there is need or better abstraction.

nicholastmosher replied:
The reason for needing this change is that previously, we had no config options that did not have a default value. Now, it should be _invalid_ to produce a `ClusterConfig` without first providing a `chart_version`. However, we do not want to put `chart_version: Option<String>` directly into our Config object because then our business logic would need  to check that it is present at every usage point.

Now, we distinguish between the Builder, which is where we collect all of our options, and the Config, which is where we are guaranteed to have all of the options we need. To get from Builder to Config you use `.build()`, which will fail if not all of the options are provided.

simlay commented on src/cluster/Cargo.toml:
```diff
@@ -11,20 +11,19 @@ description = "Tools for installing and managing Fluvio clusters"
-name = "fluvio-cluster"
-path = "src/main.rs"
-required-features = ["cli"]
+#[[bin]]
```
delete?
